### PR TITLE
Fix CCE when a type reference is used for stream types

### LINKED ISSUE #38100: [Bug]: CCE when a type reference is used for stream types
### Description

```bal
type MyIntStream stream<int, error?>;

type MYStreamUnion MyIntStream|stream<string>;

public function main() {
    var intGenObj = object {
        int number = 0;
        public isolated function next() returns record {|int value;|}|error? {
            self.number += 1;
            return {value: self.number};
        }
    };

    MyIntStream c = new(intGenObj);
}
```

Gives

```
[2022-10-06 17:50:31,897] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.semantics.model.types.BTypeReferenceType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BStreamType (org.wso2.ballerinalang.compiler.semantics.model.types.BTypeReferenceType and org.wso2.ballerinalang.compiler.semantics.model.types.BStreamType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.types.BTypeReferenceType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BStreamType (org.wso2.ballerinalang.compiler.semantics.model.types.BTypeReferenceType and org.wso2.ballerinalang.compiler.semantics.model.types.BStreamType are in unnamed module of loader 'app')
        at org.wso2.ballerinalang.compiler.desugar.Desugar.desugarStreamTypeInit(Desugar.java:6903)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:6811)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangTypeInit.accept(BLangTypeInit.java:65)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewriteExpr(Desugar.java:8873)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1474)
        at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:54)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8848)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1610)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:50)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8848)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8886)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewriteStmt(Desugar.java:8892)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1377)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:60)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8848)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:1423)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:76)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8848)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.visit(Desugar.java:829)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:174)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.rewrite(Desugar.java:8848)
        at org.wso2.ballerinalang.compiler.desugar.Desugar.perform(Desugar.java:471)
        at io.ballerina.projects.internal.CompilerPhaseRunner.desugar(CompilerPhaseRunner.java:216)
```

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.3.0-stage

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$title

Fixes #38100

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal [hunk 1]
```diff
@@ -621,6 +621,22 @@ function testCastingFromSuperStreamType() {
...
+        public isolated function next() returns record {|int value;|}? {
+            self.number += 1;
+            return {value: self.number};
+        }
+    };
+
+    MyIntStream strm = new (intGenObj);
+    assertValueEquality(1, getGenericRecordValue(strm.next()));
+    assertValueEquality(2, getGenericRecordValue(strm.next()));
+}
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal:638]: can you add a test with empty initializer
```ballerina
MyIntStream strm = new ();
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal [hunk 2]
```diff
@@ -621,6 +621,22 @@ function testCastingFromSuperStreamType() {
     var _ = <stream<int, error?>>localEvenNumberStream;
 }
 
+type MyIntStream stream<int>;
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal:624]: Add a test with the object type being a reference type. https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L4216 will have to change as well
eg:
```ballerina
public type ObjectType object {
    public isolated function next() returns record {|int value;|}?;
};

function dummyTest() {
      ObjectType intGenObj = object {
            int number = 0;
            public isolated function next() returns record {|int value;|}? {
                self.number += 1;
                return {value: self.number};
            }
        };

        MyIntStream strm = new (intGenObj);

        print(strm.next());
}
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal (hunk 1 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal:657]: Added in [e23a1ba](https://github.com/ballerina-platform/ballerina-lang/pull/38101/commits/e23a1ba0ce174af8153772f5a75c0aa04682edd1)

File: tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal (hunk 2 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/stream/stream-value.bal:624]: Added in [e23a1ba](https://github.com/ballerina-platform/ballerina-lang/pull/38101/commits/e23a1ba0ce174af8153772f5a75c0aa04682edd1)

**Review by pcnfernando — APPROVED** [review]

**Comment by pcnfernando** [comment]
Can we port the fix to master?
